codeU2.py

```diff
@@ -0,0 +1,210 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Jan 13 10:22:28 2017
+Exercise set 2, Code U
+@author: Luiza
+"""
+import logging
+import unittest
+
+def dfs(tree, start):
+
+    """Depth-first search
+    """
+
+    if tree is None:
+        return
+    visited = set()
+    stack = [start]
+    while stack:
+        vertex = stack.pop()
+        if not vertex:
+            continue
+        visited.add(vertex)
+        stack.extend(set(tree[vertex]) - visited)
+    return visited
+
+def dfs_path(tree, start, end, path=None):
+
+    """Returns all full paths from start node
```

Comment:
The function works for the cases, when the input is a graph as well or the element indeed occurs multiple times, that is why multiple paths are assumed. Added the comment :)